### PR TITLE
Add mitra embeddings support example

### PR DESCRIPTION
**Pull request by Oceanusity**
## *Issue #, if available:*

*Description of changes:*
This PR adds the provides a tool named gen_mitra_embeddings.py to automatically generate the node features for target nodes. Note that current Mitra model supports the number of labels should be no more than 10.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tools/gen_mitra_embedding.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'python'))
+
+from autogluon.tabular import TabularPredictor
+from autogluon.tabular.models.mitra._internal.data.dataset_finetune import DatasetFinetune
+from autogluon.tabular.models.mitra._internal.config.enums import Task, LossName, ModelName
+from autogluon.tabular.models.mitra._internal.core.trainer_finetune import CollatorWithPadding
+
+from graphstorm.data.ogbn_datasets import OGBTextFeatDataset
+from graphstorm.data.movielens import MovieLens100kNCDataset
+from graphstorm.data.dataset import ConstructedGraphDataset
```
[the remark this answers is not in the document]
Comment (Oceanusity) [reply tools/gen_mitra_embedding.py]: Got it. Will rewrite the dataset part.

File: tools/README.md [hunk 2]
```diff
@@ -118,6 +118,46 @@ The conversion script has a minimum memory requirement of 2X of the size of the
...
+### Custom Graph Example
+
+For custom graphs constructed from parquet files using gconstruct:
+
+```bash
+# First, construct the DGL graph from parquet files
+python3 -m graphstorm.gconstruct.construct_graph \
+    --conf-file config.json \
+    --output-dir /tmp/constructed \
+    --graph-name my-graph
```
Comment (jalencato) [inline tools/README.md]: Shall the Mitra embedding be done before the graph construction?

Comment (Oceanusity) [reply tools/README.md]: Guessing it will be the same as conducting construction first and then extract the embedding? If the final target is to train the GNN.

File: tools/README.md [hunk 3]
```diff
@@ -118,6 +118,60 @@ The conversion script has a minimum memory requirement of 2X of the size of the
 python3 convert_feat_to_wholegraph.py --dataset-path ogbn-mag240m-2p --node-feat-names paper:feat --low-mem
 ```
 
+## Generate Mitra embeddings for graph data
```
Comment (jalencato) [inline tools/README.md]: Missing dependencies installation guide.

File: tools/README.md [hunk 4]
```diff
@@ -118,6 +118,99 @@ The conversion script has a minimum memory requirement of 2X of the size of the
...
+GraphStorm provides a tool to generate Mitra embeddings using AutoGluon's Mitra TFM (Tabular Foundation Model). The `gen_mitra_embedding.py` script supports both the MovieLens-100k dataset and custom parquet data. Mitra embeddings can be used as pre-trained node features to improve model performance.
```
Comment (zhjwy9343) [inline tools/README.md]: Could you add the URL of Mitra TFM here to give users a reference.

File: tools/README.md [hunk 5]
```diff
@@ -118,6 +118,99 @@ The conversion script has a minimum memory requirement of 2X of the size of the
...
+
+**Important**: The current version of Mitra TFM supports a maximum of 10 classes for multiclass classification problems.
```
Comment (zhjwy9343) [inline tools/README.md]: Could you specify which version of Mitra TFM, e.g., v1.*, is the `current` version?

File: tools/README.md [hunk 6]
```diff
@@ -118,6 +118,99 @@ The conversion script has a minimum memory requirement of 2X of the size of the
...
+# Output: output/user/mitra_embeddings.parquet
+python3 tools/gen_mitra_embedding.py --dataset movie-len \
+                            --raw-data-path data/ml-100k/ml-100k \
+                            --dataset_path data/ml-100k-parquet/ \
+                            --savedir output \
+                            --target-ntype user \
+                            --label-name gender \
+```
+
+**MovieLens Parameters:**
```
Comment (zhjwy9343) [inline tools/README.md]:
```suggestion
**Parameters:**
```

File: tools/README.md [hunk 7]
```diff
@@ -118,6 +118,99 @@ The conversion script has a minimum memory requirement of 2X of the size of the
...
+For `--target-ntype user`:
+- `gender`: Binary classification (0=Female, 1=Male)
+- `age`: Age group classification (0-5, representing 6 age groups)
+- `occupation`: Occupation classification (0-20, representing 21 occupations)
+
+**Note**: The `--raw-data-path` should point to the directory containing the extracted MovieLens files (u.user, u.item, u.occupation, etc.). The tool will convert this to parquet format and save to `--dataset_path`. Output embeddings will be saved to `{savedir}/{target-ntype}/`.
+
+### Custom Dataset Example
+
+For custom datasets, parquet files should be organized in subdirectories by node type:
```
Comment (zhjwy9343) [inline tools/README.md]:
```suggestion
For custom datasets, parquet files should be organized in subdirectories by node types:
```

File: tools/README.md [hunk 8]
```diff
@@ -118,6 +118,99 @@ The conversion script has a minimum memory requirement of 2X of the size of the
 python3 convert_feat_to_wholegraph.py --dataset-path ogbn-mag240m-2p --node-feat-names paper:feat --low-mem
 ```
 
+## Generate Mitra embeddings for graph data
```
Comment (zhjwy9343) [inline tools/README.md]: Could you provide the context of why use Mitra TFM here? And how Mitra TFM will process the node features briefly?

File: tools/README.md [hunk 9]
```diff
@@ -118,6 +118,99 @@ The conversion script has a minimum memory requirement of 2X of the size of the
...
+- Input parquet files: `{dataset_path}/{target-ntype}/*.parquet`
+- Output embeddings: `{savedir}/{target-ntype}/mitra_embeddings.parquet`
+
+**Output Files:**
+Output files are saved to `{savedir}/{target-ntype}/`:
+- `mitra_embeddings.parquet`: Embeddings with columns in order:
+  1. `node_id`: Node identifier
+  2. Target column (name from `--target-col` or `--label-name`): Label values
+  3. Embedding dimensions: `0`, `1`, `2`, ... (numeric column names)
+- `mitra_embeddings.pt`: PyTorch tensor of embeddings only (for backward compatibility)
```
Comment (zhjwy9343) [inline tools/README.md]: What does the `for backward compatibiltiy` mean here?

File: tools/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (Oceanusity) [reply tools/README.md]: Have added the parquet file reference: https://graphstorm.readthedocs.io/en/latest/tutorials/own-data.html. And point out `Note that for the target node type, only the `mitra_embedding` feature is required.` at the end of the section.

File: tools/download_movielens.py [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    MovieLens Dataset Download and Conversion Tool
+
+    This tool downloads the MovieLens-100k dataset and converts all data files
+    (users, movies, ratings) to parquet format for use with GraphStorm gconstruct
```
Comment (zhjwy9343) [inline tools/download_movielens.py]:
```suggestion
    (users, movies, ratings) to parquet format for the
```

File: tools/download_movielens.py [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+        return raw_dir
+    except Exception as e:
+        raise RuntimeError(f"Failed to download MovieLens 100k dataset: {e}")
+
+
+def convert_movielens_to_parquet(raw_data_path, output_dir):
+    """
+    Convert all MovieLens raw data files to parquet format.
+    
+    This function converts users, movies, and ratings data into separate
```
Comment (zhjwy9343) [inline tools/download_movielens.py]:
```suggestion
    This function converts `users`, `movies`, and `ratings` data into separate
```

File: tools/download_movielens.py [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+    parquet files organized by node/edge type for GraphStorm gconstruct.
```
Comment (zhjwy9343) [inline tools/download_movielens.py]:
```suggestion
    parquet files organized by node/edge types for GraphStorm gconstruct commands.
```

File: tools/download_movielens.py [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+    
+    Parameters
+    ----------
+    raw_data_path : str
+        Path to the directory containing MovieLens files (e.g., data/ml-100k/ml-100k)
+    output_dir : str
+        Directory where parquet files will be saved in subdirectories by type
```
Comment (zhjwy9343) [inline tools/download_movielens.py]:
```suggestion
        Directory where parquet files will be saved in subdirectories by types
```

File: tools/gen_mitra_embedding.py [hunk 14]
```diff
@@ -0,0 +1,579 @@
...
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Mitra Embedding Generation Tool for GraphStorm
+
+    This tool generates Mitra embeddings from custom datasets in parquet format
+    for use in GraphStorm.
```
Comment (zhjwy9343) [inline tools/gen_mitra_embedding.py]:
```suggestion
    to be used in GraphStorm.
```

File: tools/gen_mitra_embedding.py [hunk 15]
```diff
@@ -0,0 +1,579 @@
...
+
+    Expected Input Directory Structure
+    -----------------------------------
+    The tool expects parquet files organized by node type:
```
Comment (zhjwy9343) [inline tools/gen_mitra_embedding.py]:
```suggestion
    The tool expects parquet files organized by node types:
```

File: tools/gen_mitra_embedding.py [hunk 16]
```diff
@@ -0,0 +1,579 @@
...
+       └── target_ntype/
+           ├── part-00000.parquet
+           ├── part-00001.parquet
+           └── part-00002.parquet
+    
+    Parquet File Requirements
+    -------------------------
+    Each parquet file must contain:
+    - Feature columns: Numeric columns used for embedding generation
+    - Label column: Target column for classification (specified by --label-name)
```
Comment (zhjwy9343) [inline tools/gen_mitra_embedding.py]: Below the column identifier is `label_col`, should use it here?

File: tools/gen_mitra_embedding.py [hunk 17]
```diff
@@ -0,0 +1,579 @@
...
+        print(f"Node ID column '{node_id_col}' not found, creating sequential IDs")
+    
+    # Check if label column exists
+    if label_col not in df.columns:
+        raise ValueError(f"Label column '{label_col}' not found in data. Available columns: {list(df.columns)}")
+    
+    # Select features
+    if feature_cols is None:
+        # Use all columns except label and node_id as features
+        feature_cols = [col for col in df.columns if col not in [label_col, node_id_col]]
```
Comment (zhjwy9343) [inline tools/gen_mitra_embedding.py]: What if only `label_col` and `node_id_col` exist, but no feature column found?

File: tools/gen_mitra_embedding.py [hunk 18]
```diff
@@ -0,0 +1,579 @@
...
+        Input data to generate embeddings for
+    save_dir : str, optional
+        Directory to save intermediate embedding files. If None, uses a temporary directory
+        
+    Returns
+    -------
+    tuple of (np.ndarray, np.ndarray)
+        - embeddings: Instance embeddings from the model [N, embedding_dim]
+        - avg_embeddings: Average embeddings across features [N, embedding_dim]
+        Returns empty arrays if no embeddings are generated
```
Comment (zhjwy9343) [inline tools/gen_mitra_embedding.py]:
```suggestion
    Returns empty arrays if no embeddings are generated
```

File: tools/gen_mitra_embedding.py [hunk 19]
```diff
@@ -0,0 +1,579 @@
...
+    if self.cfg.model_name == ModelName.TABPFN:
+        pad_to_max_features = True
+    elif self.cfg.model_name in [ModelName.TAB2D, ModelName.TAB2D_COL_ROW, ModelName.TAB2D_SDPA]:
+        pad_to_max_features = False
+    else:
+        raise NotImplementedError(f"Model {self.cfg.model_name} not implemented")
+
+    loader = torch.utils.data.DataLoader(
+        dataset,
+        batch_size=1,
```
Comment (zhjwy9343) [inline tools/gen_mitra_embedding.py]: Why do we use 1 as batch size?

Comment (Oceanusity) [reply tools/gen_mitra_embedding.py]: Follow the implementation of Mitra to use 1 as the batch size to extract the support and query data.

File: tools/gen_mitra_embedding.py (hunk 17 above)
Comment (Oceanusity) [reply tools/gen_mitra_embedding.py]: fixed.

File: tools/gen_mitra_embedding.py [hunk 20]
```diff
@@ -0,0 +1,579 @@
...
+                    gc.collect()
+        
+        torch.cuda.empty_cache()
+        gc.collect()
+    
+    if len(embeddings) > 0:
+        temp_emb_file = os.path.join(save_dir, f'embeddings_part_{file_counter}.npy')
+        temp_avg_file = os.path.join(save_dir, f'avg_embeddings_part_{file_counter}.npy')
+        
+        try:
```
[the remark this answers is not in the document]
Comment (Oceanusity) [reply tools/gen_mitra_embedding.py]: Yes, the information is printed in the main function after the features is generated in the tool file. https://github.com/Oceanusity/graphstorm/blob/6679f2d325e10f9e7718052d0a6e18c33a92b184/tools/gen_mitra_embedding.py#L571
